### PR TITLE
FileSystem writeFile and writeFileSync options parameter added

### PR DESCRIPTION
**Pull request by Wikront**

### CODE REVIEW

**Comment by yichoi** [comment]
@Wikront, Would you please add [DCO](https://github.com/Samsung/iotjs/wiki/IoT.js-Developer's-Certificate-of-Origin-1.0) and update PR along with [IoT.js development process](https://github.com/Samsung/iotjs/wiki/Development-Process) again ?

File: src/js/fs.js [hunk 1]
```diff
@@ -311,11 +324,21 @@ fs.writeFile = function(path, data, callback) {
...
-fs.writeFileSync = function(path, data) {
+fs.writeFileSync = function(path, data, options) {
   checkArgString(path);
   checkArgBuffer(data);
+  var flag;
+  var opt = 'w';
+  if (util.isObject(options) && !util.isNullOrUndefined(options.flag)) {
+    opt = options.flag;
+  }
+  if (util.isObject(options) && (opt === 'a' || opt === 'w')) {
```
Comment (hs0225) [inline src/js/fs.js]: Please don't repeat `util.isObject(options)`.

**Comment by Wikront** [comment]
added test suite, code modified along suggestions

File: src/js/fs.js [hunk 2]
```diff
@@ -264,16 +264,26 @@ fs.readFileSync = function(path) {
...
   checkArgBuffer(data);
   checkArgFunction(callback);
 
   var fd;
   var len;
   var bytesWritten;
+  var flag = 'w';
+  if (util.isObject(options) && !util.isNullOrUndefined(options.flag)) {
+    var opt = options.flag;
+    if (opt === 'a' || opt === 'w') {
```
Comment (hs0225) [inline src/js/fs.js]: `flag` can be anything other than `a`and `w`. I don't think if statement is needed.

File: src/js/fs.js [hunk 3]
```diff
@@ -311,11 +321,18 @@ fs.writeFile = function(path, data, callback) {
...
 
 
-fs.writeFileSync = function(path, data) {
+fs.writeFileSync = function(path, data, options) {
   checkArgString(path);
   checkArgBuffer(data);
+  var flag = 'w';
+  if (util.isObject(options) && !util.isNullOrUndefined(options.flag)) {
+    var opt = options.flag;
+    if (opt === 'a' || opt === 'w') {
```
Comment (hs0225) [inline src/js/fs.js]: ditto.

File: test/run_pass/test_fs_writefile_flag_unlink.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  @STDOUT=Pass
+ */
+
+var fs = require('fs');
+var assert = require('assert');
+
+var file1 = "resources/tobeornottobe.txt";
+var file2 = "resources/tobeornottobe_async.txt";
+var file3 = "resources/flag.txt";
+var file4 = "resources/tobeornottobe_flag.txt";
```
Comment (hs0225) [inline test/run_pass/test_fs_writefile_flag_unlink.js]: For Nuttx, please change to the absolute path like below.
``` js
var file = process.cwd() + ...
```

File: test/run_pass/test_fs_writefile_flag_unlink.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      assert.equal(exists, true);
+      fs.readFile(file1, function(err, buf2) {
+        assert.equal(err, null);
+        assert.equal(buf1.toString(), buf2.toString());
+        fs.unlink(file2, function(err) {
+          assert.equal(err, null);
+          fs.exists(file2, function(exists) {
+            assert.equal(exists, false);
+            fs.readFile(file3, function(err, buf3){
+              fs.writeFile(file2, buf3, {flag: 'a'}, function(){
```
Comment (hs0225) [inline test/run_pass/test_fs_writefile_flag_unlink.js]: Please remove `file2` after the test. After the test is finished, no files should be created.

File: src/js/fs.js (hunk 2 above)
Comment (zherczeg) [inline src/js/fs.js]: I think the problem is that user can pass anything. We could make it:

    if (opt == 'a') {
      flag == opt;
    }

Comment (hs0225) [inline src/js/fs.js]: `flag` can be 'rs', 'r+', 'wx', 'w+', 'ax' and many more. If an invalid value is passed, `fs.open` will throw a type error.

**Comment by yichoi** [comment]
@Wikront  No further updates? rebase needed

File: src/js/fs.js [hunk 6]
```diff
@@ -264,16 +264,26 @@ fs.readFileSync = function(path) {
...
+    callback = options;
+  }
   checkArgString(path);
   checkArgFunction(callback);
 
   var fd;
   var len;
   var bytesWritten;
   var buffer = ensureBuffer(data);
+  var flag = 'w';
```
Comment (galpeter) [inline src/js/fs.js]: Maybe we could have the lines 278-284 and 328-334 move to a helper method so we can avoid code duplication.

**Comment by yichoi** [comment]
no updates for a long time

**Pull request closed without merging**